registry/repositoryClient.go

```diff
@@ -0,0 +1,163 @@
+package registry
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	artifactspec "github.com/oras-project/artifacts-spec/specs-go/v1"
+	"oras.land/oras-go/v2"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2/registry/remote"
+)
+
+const (
+	maxBlobSizeLimit     = 32 * 1024 * 1024 // 32 MiB
+	maxManifestSizeLimit = 4 * 1024 * 1024  // 4 MiB
+)
+
+// repositoryClient implements Repository
+type repositoryClient struct {
+	remote.Repository
+}
+
+// Resolve resolves a reference(tag or digest) to a manifest descriptor
+func (c *repositoryClient) Resolve(ctx context.Context, reference string) (ocispec.Descriptor, error) {
+	return c.Repository.Resolve(ctx, reference)
+}
+
+// ListSignatures returns signature manifests filtered by fn given the
+// artifact manifest descriptor
+func (c *repositoryClient) ListSignatures(ctx context.Context, desc ocispec.Descriptor, fn func(signatureManifests []ocispec.Descriptor) error) error {
+	if err := c.Repository.Referrers(ctx, ocispec.Descriptor{
+		Digest: desc.Digest,
+	}, ArtifactTypeNotation, func(referrers []artifactspec.Descriptor) error {
+		var sigManifestDesc []ocispec.Descriptor
+		for _, referrer := range referrers {
+			sigManifestDesc = append(sigManifestDesc, ocispecDescriptorFromArtifact(referrer))
+		}
+		return fn(sigManifestDesc)
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+// FetchSignatureBlob returns signature envelope blob and descriptor given
+// signature manifest descriptor
+func (c *repositoryClient) FetchSignatureBlob(ctx context.Context, desc ocispec.Descriptor) ([]byte, ocispec.Descriptor, error) {
+	sigManifest, err := c.getSignatureManifest(ctx, desc)
+	if err != nil {
+		return nil, ocispec.Descriptor{}, err
+	}
+	if len(sigManifest.Blobs) == 0 {
+		return nil, ocispec.Descriptor{}, errors.New("signature manifest missing signature envelope blob")
+	}
+	sigDesc, err := c.Repository.Blobs().Resolve(ctx, sigManifest.Blobs[0].Digest.String())
+	if err != nil {
+		return nil, ocispec.Descriptor{}, err
+	}
+	if sigDesc.Size > maxBlobSizeLimit {
+		return nil, ocispec.Descriptor{}, fmt.Errorf("signature blob too large: %d", sigDesc.Size)
+	}
+	sigBlob, err := content.FetchAll(ctx, c.Repository.Blobs(), sigDesc)
```

Comment:
why not use sigManifest.Blobs[0] directly?

Reply:
sigManifest.Blobs[0] is the descriptor of the sigBlob, we want the sigBlob envelope itself right?